plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetColumnIOConverter.java

```diff
@@ -71,26 +76,114 @@ private ParquetColumnIOConverter() {}
         }
         if (type instanceof MapType) {
             GroupColumnIO groupColumnIO = (GroupColumnIO) columnIO;
-            MapType mapType = (MapType) type;
-            GroupColumnIO keyValueColumnIO = getMapKeyValueColumn(groupColumnIO);
-            if (keyValueColumnIO.getChildrenCount() != 2) {
+            Optional<ColumnIO> columnForMapKey = getColumnForMapKey(context, groupColumnIO);
+            Optional<ColumnIO> columnForMapValue = getColumnForMapValue(context, groupColumnIO);
+            if (columnForMapKey.isEmpty() || columnForMapValue.isEmpty()) {
                 return Optional.empty();
             }
-            Optional<Field> keyField = constructField(mapType.getKeyType(), keyValueColumnIO.getChild(0));
-            Optional<Field> valueField = constructField(mapType.getValueType(), keyValueColumnIO.getChild(1));
+            Optional<Field> keyField = constructField(getMapKeyContext(context), columnForMapKey);
+            Optional<Field> valueField = constructField(getMapValueContext(context), columnForMapValue);
             return Optional.of(new GroupField(type, repetitionLevel, definitionLevel, required, ImmutableList.of(keyField, valueField)));
         }
         if (type instanceof ArrayType) {
             GroupColumnIO groupColumnIO = (GroupColumnIO) columnIO;
-            List<Type> types = type.getTypeParameters();
-            if (groupColumnIO.getChildrenCount() != 1) {
+            Optional<ColumnIO> arrayElementColumn = getColumnForArrayElement(context, groupColumnIO);
+            if (arrayElementColumn.isEmpty()) {
                 return Optional.empty();
             }
-            Optional<Field> field = constructField(types.get(0), getArrayElementColumn(groupColumnIO.getChild(0)));
+            Optional<Field> field = constructField(getArrayElementContext(context), arrayElementColumn.get());
             return Optional.of(new GroupField(type, repetitionLevel, definitionLevel, required, ImmutableList.of(field)));
         }
         PrimitiveColumnIO primitiveColumnIO = (PrimitiveColumnIO) columnIO;
         RichColumnDescriptor column = new RichColumnDescriptor(primitiveColumnIO.getColumnDescriptor(), columnIO.getType().asPrimitiveType());
         return Optional.of(new PrimitiveField(type, repetitionLevel, definitionLevel, required, column, primitiveColumnIO.getId()));
     }
+
+    protected abstract Type getType(Context context);
+
+    protected abstract Context getArrayElementContext(Context context);
+
+    protected abstract Optional<ColumnIO> getColumnForArrayElement(Context context, GroupColumnIO groupColumnIO);
```

Comment:
it could make sense. @losipiuk do you want to share your thinking before i make this change?

Reply:
+1 to simplifying this.  another option could be to make `getColumnForMapKey`, `getColumnForMapValue`, `getColumnForArrayElement` non-abstract without using context, as it's context-independent at this point.

Reply:
> it could make sense. @losipiuk do you want to share your thinking before i make this change?

I do not think it matters much too be honest. It feels like it would look a bit nicer, but given fact that this is locally used API, not something scattered through all codebase it would not make much difference. Feel free to go either way IMO.

Reply:
the code will indeed be simpler, changing

Reply:
> the code will indeed be simpler, changing

however not 100% equivalent, see comment https://github.com/trinodb/trino/pull/9104#discussion_r702812638